### PR TITLE
Handle memory arbitration trigger from non-driver execution context

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Not all the memory arbitration are triggered from driver execution context
such as async table scan prefetch or async streaming shuffle return path.
Given that, we need to allow the memory arbitration to handle such case.
To enter/leave memory arbitration, we need to skip the driver suspension
enter/leave step if it is not under driver execution context.
This PR add unit tests to trigger memory arbitration under driver init and 
parallel table build. We will fix these two cases in followup at least for parallel
table join to avoid the memory allocations during the parallel join. Otherwise
the memory arbitration will hang the arbitration process.

### CODE REVIEW

File: velox/core/QueryConfig.h [hunk 1]
```diff
@@ -205,6 +205,10 @@ class QueryConfig {
   static constexpr const char* kHashProbeFinishEarlyOnEmptyBuild =
       "hash_probe_finish_early_on_empty_build";
 
+  /// The minimum table size in row that triggers the parallel join table build.
+  static constexpr const char* kMinTableSizeForParallelJoinBuild =
```
Comment (mbasmanova) [inline velox/core/QueryConfig.h:209]: It would be nice to document this new property.

File: velox/core/QueryConfig.h [hunk 2]
```diff
@@ -205,6 +205,10 @@ class QueryConfig {
   static constexpr const char* kHashProbeFinishEarlyOnEmptyBuild =
       "hash_probe_finish_early_on_empty_build";
 
+  /// The minimum table size in row that triggers the parallel join table build.
```
Comment (mbasmanova) [inline velox/core/QueryConfig.h:208]: What does "table size in row" mean?

File: velox/exec/HashBuild.cpp [hunk 3]
```diff
@@ -766,12 +775,22 @@ bool HashBuild::finishHashBuild() {
 
       // TODO: re-enable parallel join build with spilling triggered after
       // https://github.com/facebookincubator/velox/issues/3567 is fixed.
-      const bool allowPrallelJoinBuild =
+      const bool allowParallelJoinBuild =
           !otherTables.empty() && spillPartitions.empty();
+      if (TestValue::enabled()) {
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:780]: What does this do? Maybe add a comment?

File: velox/exec/HashTable.h [hunk 4]
```diff
@@ -642,33 +630,31 @@ class HashTable : public BaseHashTable {
...
-  void pushNext(char* FOLLY_NULLABLE row, char* FOLLY_NULLABLE next);
+  void pushNext(char* row, char* next);
 
   // Finishes inserting an entry into a join hash table. If the insert
   // would fall outside of 'partitionBegin' ... 'partitionEnd', the
   // insert is not made but the row is instead added to 'overflow'.
   void buildFullProbe(
       ProbeState& state,
       uint64_t hash,
-      char* FOLLY_NULLABLE row,
```
Comment (mbasmanova) [inline velox/exec/HashTable.h:657]: Thank you for the refactoring, but it makes it harder to see material changes. Would it be possible to extract refactoring into a separate PR?

File: velox/exec/Operator.cpp [hunk 5]
```diff
@@ -415,6 +415,13 @@ void Operator::MemoryReclaimer::enterArbitration() {
   // The driver must be alive as the operator is still under memory arbitration
   // processing.
   VELOX_CHECK_NOT_NULL(driver);
+  if (FOLLY_UNLIKELY(!driver->state().isOnThread())) {
+    // NOTE: not all the memory arbitration are triggered from driver execution
```
Comment (mbasmanova) [inline velox/exec/Operator.cpp:419]: The wording here is confusing. It reads as "table scan prefetch" is an example of "driver executor" when it is the opposite.

File: velox/common/memory/tests/SharedArbitratorTest.cpp [hunk 6]
```diff
@@ -1576,6 +1589,165 @@ DEBUG_ONLY_TEST_F(
...
+
+  std::shared_ptr<core::QueryCtx> joinQueryCtx = newQueryCtx(kMemoryCapacity);
+  std::shared_ptr<core::QueryCtx> fakeQueryCtx = newQueryCtx(kMemoryCapacity);
+
+  // Set fake operator to reclaimable to allow arbitration to succeed.
+  fakeOperatorFactory_->setCanReclaim(true);
+
+  folly::EventCount waitForPrepareJoin;
+  folly::EventCount waitForFakeAllocationDone;
+  // std::atomic<bool> injectAllocationOnce{true};
```
Comment (mbasmanova) [inline velox/common/memory/tests/SharedArbitratorTest.cpp:1612]: Remove?

**Review by mbasmanova — APPROVED** [review]

**Comment by mbasmanova** [comment]
CI is red. Please, take a look.

File: velox/core/QueryConfig.h (hunk 2 above)
Comment (xiaoxmeng) [reply velox/core/QueryConfig.h:208]: Change to "The number of table rows" to make it easier to understand.

File: velox/exec/Driver.h [hunk 7]
```diff
@@ -90,7 +90,7 @@ std::ostream& operator<<(std::ostream& out, const StopReason& reason);
 // suspended, off thread or blocked.
 struct ThreadState {
   // The thread currently running this.
-  std::atomic<std::thread::id> thread{};
+  std::atomic<std::thread::id> thread{std::thread::id()};
```
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/Driver.h:93]: Yeah, they are synced. When clear the thread in driver state, we set tid to zero and thread to std::thread::id().

File: velox/exec/HashBuild.cpp (hunk 3 above)
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp:780]: Commented. Thanks!
